content/blog/2019-11-18-edge-architecture.md

```diff
@@ -0,0 +1,72 @@
+
+### Spatial Path Scheduling (SPS)
+
+The main idea of this algorithm is that, for each instruction, it calculates the cost for each possible position in the ALU array. Then the final placement cost is the minimal cost among all positions. Unlike GRST, the initial set contains the instructions having no or **at least one** scheduled inputs. Then after the calculation of the placement cost, the instruction with the highest placement cost will be scheduled first. This idea is similar to GRST, where the instructions in the critical paths will be scheduled first. To further improve the algorithm, the authors get the idea for other heuristics by comparing the results generated by SPS and simulated annealing.
+
+#### Contention Modeling
+
+The authors model two types of contention, which are ALU and link contention, respectively. For the first one, the idea is to solve the resource conflict when scheduling two instructions with the same ALU (i.e., execution unit). For ALU contention, we can further categorize it into two types. One is intra-block contention, and the other one is inter-block contention. For intra-block contention, it is similar to GRST, where the algorithm checks whether the two instructions that mapped to the same ALU may have resource conflict or not. If that is the case, then we add the penalty to the placement cost of the instruction we are scheduling. 
+
+For inter-block contention, it is not handled by GRST. The idea is that the two instructions inside two different blocks may have resource conflicts. The authors try to add the number of consumers of the already scheduled instructions to the placement cost. However, this is too conservative since the consumers of the scheduled instructions do not necessarily have resource conflicts with the instruction that will be scheduled. To counter this, the algorithm removes candidates that must not have resource conflicts. For link contention, it is not trivial to calculate because the network utilization is unknown until runtime. 
+
+To calculate the final penalty, the algorithm sums up both intra- and intro-contention for ALUs. In addition, it introduces two factors, which are fullness and criticality, respectively. For the former, it corresponds to how full a block it. For the latter, it compares the maximum path length of an instruction with the length of the critical path. The final placement cost is just a combination of all the above numbers.
+
+#### Global Register Prioritization
+
+The main idea of this heuristic is performing loop-related optimizations. The algorithm performs the following three heuristics. First, smaller loops (in terms of the number of blocks) will be scheduled first. Second, loop-carried dependencies will be scheduled first. Third, when calculating the longest path, the algorithm will also consider the predecessor and the successor of the current block.
```

Comment:
I'm a little confused by the "first" and "second" terminology here—maybe use "earlier" or something if you don't literally mean "first"?